common/persistence/sql/sqlplugin/postgres/plugin.go

```diff
@@ -90,32 +86,60 @@ func (d *plugin) createDBConnection(cfg *config.SQL) (*sqlx.DB, error) {
 		return nil, fmt.Errorf("invalid connect address, it must be in host:port format, %v, err: %v", cfg.ConnectAddr, err)
 	}
 
+	db, err := sqlx.Connect(PluginName, buildDSN(cfg, host, port, sslParams))
+	if err != nil {
+		return nil, err
+	}
+	if cfg.MaxConns > 0 {
+		db.SetMaxOpenConns(cfg.MaxConns)
+	}
+	if cfg.MaxIdleConns > 0 {
+		db.SetMaxIdleConns(cfg.MaxIdleConns)
+	}
+	if cfg.MaxConnLifetime > 0 {
+		db.SetConnMaxLifetime(cfg.MaxConnLifetime)
+	}
+
+	// Maps struct names in CamelCase to snake without need for db struct tags.
+	db.MapperFunc(strcase.ToSnake)
+	return db, nil
+}
+
+func buildDSN(cfg *config.SQL, host string, port string, sslParams url.Values) string {
 	dbName := cfg.DatabaseName
 	//NOTE: postgres doesn't allow to connect with empty dbName, the admin dbName is "postgres"
 	if dbName == "" {
 		dbName = "postgres"
 	}
-	password := ""
-	if cfg.Password != "" {
-		password = fmt.Sprintf(dataSourceNamePostgresPassword, cfg.Password)
-	}
-	db, err := sqlx.Connect(PluginName, fmt.Sprintf(dataSourceNamePostgres, cfg.User, host, port, dbName, password))
 
-	if err != nil {
-		return nil, err
+	credentialString := generateCredentialString(cfg.User, cfg.Password)
+	dsn := fmt.Sprintf(dsnFmt, credentialString, host, port, dbName)
+	if attrs := sslParams.Encode(); attrs != "" {
+		dsn += "?" + attrs
 	}
+	return dsn
+}
 
-	// Maps struct names in CamelCase to snake without need for db struct tags.
-	db.MapperFunc(strcase.ToSnake)
-	return db, nil
+func generateCredentialString(user string, password string) string {
+	userPass := user
+	if password != "" {
+		userPass += ":" + password
+	}
+	return userPass
 }
 
-// TODO: implement postgres specific support for TLS
-func registerTLSConfig(cfg *config.SQL) error {
-	if cfg.TLS == nil || !cfg.TLS.Enabled {
-		return nil
+func registerTLSConfig(cfg *config.SQL) (sslParams url.Values, err error) {
+	sslParams = url.Values{}
+	if cfg.TLS != nil && cfg.TLS.Enabled {
+		sslParams.Set("ssl", "true")
+		sslParams.Set("sslmode", "require")
```

Comment:
I know I'm late to the party, but is it wise to hard code this? How would I connect using the more secure modes `verify-ca` or `verify-full` for example?

Could as well let the user choose?
```go
sslParams.Set("sslmode", cfg.TLS.SSLMode)
```

Reply:
Hey @JosefWN you are not late! Apparently I don't have enough expertise/experience about security so I didn't think too much about that. 
Is value of `sslmode` the only thing needed to change to support all other modes like you mention? If so then I could make a PR out for this. 

Or it would be nice if you can make a PR for this. Because I am not sure how to test it.

Reply:
I think making the `sslmode` configurable is sufficient (which would also make `cfg.TLS.Enabled` redundant).

Perhaps something like this:
```go
	sslParams.Set("sslmode", cfg.SSLMode)
	if cfg.SSLMode != "disable" && cfg.TLS != nil {
		sslParams.Set("sslrootcert", cfg.TLS.CaFile)
		sslParams.Set("sslkey", cfg.TLS.KeyFile)
		sslParams.Set("sslcert", cfg.TLS.CertFile)
	}
```

The default is `prefer`, so it should be backwards compatible even if `SSLMode` is not set (but would need to test that it actually works).

Reply:
Thanks a lot! PR out here: https://github.com/uber/cadence/pull/3787/files
But I think we have to make default as require now...since people started to use it. I believe it's okay, lmk if I am wrong.